### ISSUE TITLE
Cleaning with Exceptions

### ISSUE THREAD

**Issue by Loki-L1130** [bleachbit/bleachbit#1329]
Currently, In CCleaner it is possible to exclude cookies while cleaning which is helpful a lot in the case where Cookies having Login or Session details is excluded and clean other things. 

So this is great so that it clean all cookies resulting from Casual Web Browsing but keeps Login which are essential such as Email and other work related Applications.

I think that implementing this would be great.

**Comment by Tarek-Hasan** [comment]
Hi, do you have any plan to add cookie whitelisting?
[Tron](https://www.reddit.com/r/TronScript/wiki/index/) project stopped using BleachBit for lack of this feature.

> - Remove BleachBit. This is despite the fact that I prefer BleachBit over CCleaner. Reason 1) It performs the same function as CCleaner. Reason 2) It doesn't support excluding certain cookie domains from wiping (the main reason). Once BleachBit supports cookie whitelisting, we will switch over to it exclusively and retire CCleaner from Tron

Source: https://old.reddit.com/r/TronScript/comments/ezcdc9/tron_v1100_20200205_automatically_preserve_login/

**Comment by az0** [comment]
Yes, I will add that, but it's not on the short list yet. Tell me more about the requirements for how this feature should work. For example, do you want to delete all except listed cookies, keep all except listed cookies, or both options? Do you need to manage cookies at the domain level like example.com or subdomain level like apps.example.com? Do you need to manage different cookies for the same hostname? Can the same cookie settings be used across all browsers? To start, would supporting Chrome be sufficient?

**Comment by az0** [comment]
I coded a GUI for demonstration and to get feedback. This is meant to be a  MVP with minimal functionality for its first release.  The data is fake. This demo doesn't have the backend, so it's not connected to anything, and it doesn't "work" yet.

How is this for the first release?

https://github.com/user-attachments/assets/c38272c2-3dae-436c-8423-620f7edc107b

The source code is in the [next-gen GUI repository](https://github.com/bleachbit/bleachbit_gui_next_gen).

**Comment by Coacher** [comment]
Great to see some progress on this feature. Much awaited!

**Comment by vocatus** [comment]
I'm the author of [Tron](https://github.com/bmrf/tron/) and would _love_ to see this implemented, so we can finally move away from CCleaner.

The ideal setup would be support for a plain text file with basic wildcards (`*`, `.`, etc) listing domains to exclude from cookie cleaning. 

For example, this is the [current Tron cookie whitelist](https://github.com/bmrf/tron/blob/master/resources/stage_1_tempclean/ccleaner/tron_ccleaner_cookie_whiteliest.txt). ([CCleaner `.ini` version](https://github.com/bmrf/tron/blob/master/resources/stage_1_tempclean/ccleaner/ccleaner_skip_cookie_cleanup.ini#L209))

**Comment by az0** [comment]
Who is ready to try this out? Download the special build [5.0.2.3119-cookie](https://ci.bleachbit.org/?prefix=dl/5.0.2.3119-cookie/). It is like version 5.0.2 plus a few changes, mostly the new cookie manager. If you need source (e.g., for Linux), use the cookie branch that is on GitHub.

To set cookie preferences, go to application icon - preferences - allowlist tab - click the cookie button. 

This works on Google Chrome and Firefox. It can be turned on for any related browser with a one-line change in the .xml file.

It works only on the traditional HTTP cookies, not Chrome's `IndexDB` directory or Firefox's `storage/default/http*`.

Before cleaning them, you may want to make a backup of your cookies. In Firefox, it's `cookies.sqlite`. In Chrome, it's simply `Cookies` in the profile directory.

> The ideal setup would be support for a plain text file with basic wildcards (*, ., etc) listing domains to exclude from cookie cleaning.

The new BleachBit cookie settings are written to a new JSON file in the BleachBit configuration directory. 

> For example, this is the [current Tron cookie whitelist](https://github.com/bmrf/tron/blob/master/resources/stage_1_tempclean/ccleaner/tron_ccleaner_cookie_whiteliest.txt). ([CCleaner .ini version](https://github.com/bmrf/tron/blob/master/resources/stage_1_tempclean/ccleaner/ccleaner_skip_cookie_cleanup.ini#L209))

If those hostnames don't change often (Git shows the file was last modified five years ago), you could do a one-time conversion to JSON. 

I assume you use those as default values with a portable application? I haven't tested it, but it should work by placing the JSON file in the same directory as `bleachbit.ini`. For portable mode, that's the same directory as `bleachbit.exe` too.

**Comment by az0** [comment]
The new release adds the cookie manager to the context menu, so this is a second way to access it (in addition to the preferences dialog)

Windows user: download [5.0.2.3131-cookie](https://ci.bleachbit.org/?prefix=dl/5.0.2.3131-cookie/) 

How to run on Linux

```sh
# Create a temporary directory in the home folder
mkdir -p ~/bleachbit-testing
cd ~/bleachbit-testing

# Clean up any previous clone, if any
rm -rf bleachbit

# Clone the repository
git clone https://github.com/bleachbit/bleachbit.git
cd bleachbit

# Switch to the branch with the new cookie manager feature
git checkout -b cookie origin/cookie

# Run the application
python3 bleachbit.py
```

Once you have it installed, here it is, step by step.
1. Open this new version BleachBit
2. In the left panel, navigate to the cookie option under _Google Chrome_ or _Firefox_
3. Right click on _Cookie_
4. Click on _Cookie Manager_
5. Choose your cookies and close the dialog
6. Clean cookies under Firefox or Google Chrome

**Comment by briwnvygnn** [comment]
I have extensively tested this (release 5.0.2.3131-cookie) on Chrome using about 30 sites I have whitelisted using CCleaner.

This works flawlessly - I am ready to ditch CCleaner (finally!) for BleachBit.

The thread above indicates "It can be turned on for any related browser with a one-line change in the .xml file.".  How exactly do I do this?  I use Brave and MS Edge.

I can't wait to start using this!

**Comment by az0** [comment]
> I have extensively tested this (release 5.0.2.3131-cookie) on Chrome using about 30 sites I have whitelisted using CCleaner.
> 
> This works flawlessly - I am ready to ditch CCleaner (finally!) for BleachBit.

Thank you for the feedback, and even more, I am glad it's working!
 
> The thread above indicates "It can be turned on for any related browser with a one-line change in the .xml file.". How exactly do I do this? I use Brave and MS Edge.

[This is a commit](https://github.com/bleachbit/bleachbit/commit/ba416b1520024650e7984a22c11784c47a27b2a0) that enabled it for 10 browsers, including Brave and Edge. Recent builds on the master branch, such as [5.0.2.3156-master](https://ci.bleachbit.org/?prefix=dl/5.0.2.3156-master/), include this change, so you do not need to make additional changes.

**Comment by KremmenUK** [comment]
I've also just run this through and all saved cookies still present 

I've now deleted CCleaner portable 6.39 as the popups were getting horrendous and with version 7 being slated as the worst version yet, and no V7 portable I'm now totally Bleachbit

Thanks :)

**Comment by az0** [comment]
> I've also just run this through and all saved cookies still present

Thank you for confirming the cookie manager is working! That's great to hear. I'll close this ticket now.

> I've now deleted CCleaner portable 6.39 as the popups were getting horrendous and with version 7 being slated as the worst version yet, and no V7 portable I'm now totally Bleachbit

CCleaner version 7 has been a great promotion for BleachBit!
